site/content/api/Tenant-API.md

```diff
@@ -252,7 +252,7 @@ The table below provides an overview of the standard members defined for the JSO
 | :------------------------| :-------: | :--------- | :------------ | :---------- |
 | *tenant-id*              | *yes*     | *string*   |               | The ID of the tenant. |
 | *enabled*                | *no*      | *boolean*  | `true`        | If set to `false` the tenant is currently disabled. Protocol adapters MUST NOT allow devices of a disabled tenant to connect and MUST NOT accept data published by such devices.
-| *adapters*   | *yes*     | *array*    |               | A list of configuration options valid for certain adapters only. See [Adapter Configuration Format]({{< relref "#adapter-configuration-format" >}}) for details. **NB** This array must contain at least one element - an empty array is handled as an error. **NB** Only a single entry per *type* is allowed. If multiple entries for the same type are present it is handled as an error.
+| *adapters*               | *no*      | *array*    | empty array   | A list of configuration options valid for certain adapters only. See [Adapter Configuration Format]({{< relref "#adapter-configuration-format" >}}) for details. **NB** Only a single entry per *type* is allowed. If multiple entries for the same type are present it is handled as an error.
```

Comment:
:+1:

Reply:
If this element is not mandatory, then it may also be omitted completely, right? I just want to make sure that we are all on the same page and accept that we will need to be prepared to handle the case where the element is missing ...

Reply:
Yup that is the idea. To quickly create a new tenant with default configuration (no adapters allowed) one can just create the tenant with tenant-id property

Reply:
That makes sense to me.

Maybe we should try to approach this from a usability angle. What will be the most common case when creating a new tenant? Will a new tenant usually be allowed to use all adapters or is the standard case that a tenant is added and is only allowed to use a single (or a few) adapters only? Will the default config for those adapters be acceptable in most cases or will the configuration need to be adapted in almost all cases?

I think in most practical use cases a new tenant will probably be allowed to use a certain subset of the adapters and this adapter will also be configured specifically to the request of the tenant, e.g. if devices are required to authenticate or with a root CA certificate used to authenticate the devices etc. This is the reason why we add the Tenant API in the first place, isn't it? So I guess allowing a client to leave out the *adapters* element and interpreting it as *all adapters allowed, use the adapters' default configuration* makes sense and doesn't pose a huge risk for mis-configuration. WDYT?

Reply:
Yes, I think the same. And this would not be a huge risk for mis-configuration.

But I find it not so easy to tell what the most practical use cases for new tenants will be: 
- will they almost always be fully configured (so each desired protocol adapter is configured)
- or will they often start with the default configurations of all protocol adapters?

(actually, just a rephrase of your question, @sophokles73 )

I can imagine that there might be the desire to configure the defaults for protocol adapters differently, meaning that we do not have *enabled* always `true` (or `false`), but instead depending on the adapter type.
But for now this only makes things more complicated, and what if in the end this is **not** the standard case?

So my opinion is: let leave it simple as we currently think: 

missing adapters element -> all protocol adapters enabled, use the default configuration.

Reply:
These are some interesting points. So far I assumed, that for each adapter a tenant should be allowed to use, it must be configured in the `adapters` array. See also the note in the `Adapter Configuration Format` section:
> If for a certain protocol adapter no configuration is present it has the same semantic as a disabled protocol adapter with default values.

This has the disadvantage that for a new tenant a minimal adapter configuration [which consists atm just of the `adapter-id`] must be provided.

A default "enabled-all" semantic  could lead to problem as soon as we have protocol adapters which will result in a high overhead (e.g. memory consumption) per tenant. To solve this we would need different default flags per protocol adapter as @sysexcontrol suggested.

As for now, we don't have such protocol adapters like that, so I would be fine with a changed semantic in that way too.

Should I update the regarding section, so that when no adatper configuration is present an enabled adapter with default values is considered?

Reply:
FMPOV: yes, please :-)

Reply:
I updated the note in the 'Adapter Configuration Format' section:

Reply:
:+1: that results in a clear concept:
no config present -> defaults apply. No special cases!

If we find problems with that concept later on, we can adopt/extend it then, but not now.